DOCSP-23027: BsonExtraElements Annotation

ccho-mongodb opened the pull request:
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-23027
Staging - https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-23027-bsonextraelements-example/fundamentals/data-formats/pojo-customization/

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [x] Are all the links working?

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -364,6 +374,54 @@ The annotations in the example POJO specify the following behavior:
+:ref:`previous example <bson-annotation-code-example>` to accommodate the
+following changes to your data schema:
+
+- The ``name`` field is no longer needed, but must the data must remain for
```
but must -> but

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -364,6 +374,54 @@ The annotations in the example POJO specify the following behavior:
+field mapping. Consider a situation in which your application needs to handle
+data that conforms to different data schema versions. When your POJO fields
+change to accommodate the schema, you can continue to read, but not
+update, the remainder of the MongoDB document data from the field that uses
```
I'm not sure what you mean that you can read but not update the remainder of the data.  It seems like you can both read _and_ update.

ccho-mongodb replied:
To clarify, you're referring to generally being able to update the data and not when using a POJO field with the BsonExtraElements annotation? When reading the [description of the annotation](https://jira.mongodb.org/browse/DOCSP-22048), I inferred that the BsonExtraElements annotation would decode to a POJO, but not encode the field's value and pass it back to MongoDB.

I think while rewriting this section, I placed this info in a spot that made it confusing as to whether it was a requirement of the hypothetical situation or general capabilities of using a POJO to work with MongoDB data.

ccho-mongodb replied:
From reading your next comment, it sounds like you can update the unstructured part and that the field data gets encoded in an update operation. Will test it out and update the descriptions accordingly.

ccho-mongodb replied:
Wasn't able to update the MongoDB data using a POJO field marked with the annotation, so I'm assuming you meant you could update POJO fields generally, but not with the BsonExtraElements annotation.

jyemin replied:
I just tried it and it worked as I expected.  Can you share your reproducer?

ccho-mongodb replied:
https://gist.github.com/ccho-mongodb/98a7a6839c2b7627fb0cc2637151b2fa

I was able to set the "material" field on the "metadata" subdocument (annotated with @BsonExtraElements) using a POJO only on insert. Any attempt to update that field using the POJO (shown in the gist) resulted in no changes.

Isn't that the expected behavior? If not, what am I doing incorrectly?

jyemin replied:
OK, I think I see the problem.  The `metadata` field does not result in a sub-document.  The name of a field annotated with `BsonExtraElements` is actually ignored, and the fields within the document itself are added at the same level as the other fields.  So when you insert the `NewProductPojo` instance in your example, you should end up with:

```
{ 
    _id: "MDB0123",
   material: "iron"
}
```

and not

```
{ 
    _id: "MDB0123",
   metadata: { material: "iron"}
}
```

Does that clear it up?

ccho-mongodb replied:
Yes, from your explanation, I think my original understanding matched what you're saying, but needed to be more specific in the wording of the explanation in the docs.

To summarize, you can change the underlying data, but not from the BsonExtraElements-annotated field since it's only decoded and not encoded. E.g. I can update the MongoDB document by directly updating the "material" field, but not from the "metadata" POJO field.

jyemin replied:
I don't think what you're saying is correct.  It _is_ encoded, and either insertOne or replaceOne will write the context of the field annotated with BsonExtraElements into the top-level document.  

We can touch base tomorrow to clarify

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -364,6 +374,54 @@ The annotations in the example POJO specify the following behavior:
+update, the remainder of the MongoDB document data from the field that uses
+this annotation.
+
+Consider a situation in which you need to update the ``Product`` POJO from the
```
This is a great example, and actually one I hadn't thought of.  I think the more common case though is where you have a semi-structured schema where you want to model the structured part as POJO fields while still being able to read/update the unstructured part as a Document.  Perhaps we could add an example for that as well, or at least mention the use case?

ccho-mongodb replied:
Based on our conversation, I'm planning to change the situation to talk about a situation in which you have a partially-structured schema.

I'd like any input you have on which example you think would be most suitable:

1) Once the BsonExtraElements field is established in the POJO, update the document and application's usage of that data without having to change the serial version id. E.g. Add a "hidden: true" field/value to the document to instruct the application from surfacing that POJO from an API call.

2) Lazy data migration: the goal is to get all the data to resemble the new schema reflected in the POJO, but it's updated only when you access the document rather than in bulk. Not common, but have encountered this situation before.

3) Overload application models to contain related metadata that other processes generate (such as search weights/relevance and ML data). Doesn't provide a very strong case for using the field unless the application needed to surface that data.

ccho-mongodb commented:
@jyemin I fixed the description and updated the example, please take a look when you get the chance.

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -300,14 +300,18 @@ package:
        - `BsonDocument <{+api+}/apidocs/bson/org/bson/BsonDocument.html>`__
        - Map<String, Object>
 
+       When serializing a POJO through an insert or replace operation, the
```
I don't think this is true.  Rather, I think each field in the Map is stored at the same level as the other fields in the POJO.

ccho-mongodb replied:
Sorry, I was focused on updating the other part and forgot to update this mention.

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -382,45 +386,57 @@ BsonExtraElements Example
+     "name": "MDB0123",
+     "serialNumber": "62e2...",
+     "relatedItems": null,
+     "additionalInfo": {
```
This is not how it's supposed to work.  Rather, it should be:

```
   {
     "name": "MDB0123",
     "serialNumber": "62e2...",
     "relatedItems": null,
     "dimensions": "3x4x5",
     "weight": "256g"
   }
```

See [this test case](https://github.com/mongodb/mongo-java-driver/blob/95cb47a61f2896044e69ab73d59306e1527686e2/bson/src/test/unit/org/bson/codecs/pojo/PojoRoundTripTest.java#L528-L531) in the driver.

Do you have an example application that shows it working this way?

ccho-mongodb replied:
Sorry, I think the disconnect is that I was portraying what was in the POJO, not MongoDB since the latter is redundant. I'll show the output using a generic toString() method to make it more apparent since that would show the Java class names.

ccho-mongodb replied:
E.g.
```
ProductWithBsonExtraElements [name=MDB0123, serialNumber=62e2b9c5c32cb23e8cb0c95e, relatedItems=null, additionalInfo=Document{{dimensions=5x5x5}}]
```

ccho-mongodb commented:
Hey @jyemin, I made updates to remove the content in the annotation table and updated the description and output to make it more clear what the POJO contains. I also added my sample code for testing purposes only: https://github.com/mongodb/docs-java/pull/242/files#diff-295533015998e6eb3072ac03e18e8fe140c7c2fc29b6b6592b82fb87bb2549d5

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static CodecRegistry getCodecRegistry() {
+        ClassModel<ProductWithBsonExtraElements> classModel = ClassModel.builder(ProductWithBsonExtraElements.class).
+                conventions(Arrays.asList(Conventions.ANNOTATION_CONVENTION)).build();
```
I think this convention is enabled by default.  Any reason to include it here explicitly?  i.e. you can just do

    PojoCodecProvider.builder().automatic(true).build();

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+
+public class BsonExtraElementsExample {
+
+	private static String DOC_NAME_VALUE = "MDB0123";
```
Nit: fix indent

jyemin replied:
Nit: make it final

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+    	collection.drop();
+
+    	ProductWithBsonExtraElements doc = new ProductWithBsonExtraElements(DOC_NAME_VALUE);
+    	collection.insertOne((T) doc);
```
If you do the above, this (unsafe) cast is redundant and should be removed.

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    private static <T> void setup(MongoCollection<T> collection) {
```
Don't make this generic.  Instead just take an instance of `MongoCollection<ProductWithBsonExtraElements>`

ccho-mongodb replied:
I did this intentionally so other writers could use this class as a model for creating other small applications to test their code examples. Are there any disadvantages in using generic types for this use case?

jyemin replied:
Yes, because the cast to `T` is unsafe and generates a warning.

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+
+    // Update without the POJO; match the annotations for field names
+    private static void addOtherFields(MongoClient client) {
+    	MongoCollection collection = client.getDatabase("sample_pojo").getCollection("ProductsWithMoreInfo");
```
This should be of type `MongoCollection<Document>`

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/ProductWithBsonExtraElements.java:
```diff
@@ -0,0 +1,75 @@
+        this.name = name;
+    }
+
+	public String getName() {
```
The indents in this file should be made consistent

jyemin replied:
The indentation is still inconsistent.

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/ProductWithBsonExtraElements.java:
```diff
@@ -0,0 +1,75 @@
+    @BsonRepresentation(BsonType.OBJECT_ID)
+    private String serialNumber;
+
+    @BsonIgnore
```
Does this property serve any purpose in the example?  If not, it should be removed.

ccho-mongodb replied:
The context of the example mentions an addition to the ProductPojo, so keeping everything else was implied. It's not strictly necessary, but would likely make more sense to the next writer.

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/ProductWithBsonExtraElements.java:
```diff
@@ -0,0 +1,75 @@
+import org.bson.codecs.pojo.annotations.BsonProperty;
+import org.bson.codecs.pojo.annotations.BsonRepresentation;
+
+@BsonDiscriminator(value="ProductWithBsonExtraElements", key="_cls")
```
Is this annotation necessary the example?  If not, it should be removed

ccho-mongodb replied:
Similar reason as mentioned in this comment https://github.com/mongodb/docs-java/pull/242#discussion_r939022903

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private static <T> void printDocuments(MongoCollection<T> collection) {
+    	 List<T> products = new ArrayList<T>();
```
Can be shortened to just:

    collection.find().into(new ArrayList<>()).forEach(System.out::println);

ccho-mongodb replied:
Is there any disadvantage to keeping it as is? I was thinking that for writer testing purposes, keeping it as a lambda could allow them to insert debugging/testing logic more easily.

jyemin replied:
Just less idiomatic Java but it's ok.

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -419,24 +415,19 @@ code example:
 additional fields that describe the products such as ``dimensions`` and
-``weight``, a document retrieved using the ``Product`` POJO might resemble
-the following JSON object:
+``weight``, the document retrieved using the ``Product`` POJO might resemble
```
I still feel like it would aid clarity if you show the JSON exactly instead of just describing it.  The way it is, there is room for misinterpretation.

ccho-mongodb commented:
@jyemin Thanks for all the feedback, ready for another review.

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,70 @@
+	private static CodecRegistry getCodecRegistry() {
+		PojoCodecProvider pojoCodecProvider = PojoCodecProvider.builder().automatic(true).build();
+		return fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
+
```
Nit: extra whitespace

ccho-mongodb replied:
> The indentation is still inconsistent.

Sorry, strange conversion of tabs and spaces makes it look ok in the IDE. Need to adjust my IDE settings as mentioned in https://github.com/mongodb/docs-java/pull/242#discussion_r940305061

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,70 @@
+
+	private static CodecRegistry getCodecRegistry() {
+		PojoCodecProvider pojoCodecProvider = PojoCodecProvider.builder().automatic(true).build();
+		return fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
```
Just FYI, this can be simplified to:

    return fromProviders(getDefaultCodecRegistry(), pojoCodecProvider);

since a registry is also a provider

jyemin commented on source/includes/fundamentals/code-snippets/bson-extra-elements/BsonExtraElementsExample.java:
```diff
@@ -0,0 +1,70 @@
+
+public class BsonExtraElementsExample {
+
+	private static final String DOC_NAME_VALUE = "MDB0123";
```
Have we standardized on 8 space indents?  4 is more typical.

ccho-mongodb replied:
I think it's just a settings issue with my IDE; haven't had time to set the formatting. Will convert tabs to spaces and use 4 space indents.

jyemin reviewed (approved):
Just whitespace issues.  LGTM!
